File: worker/cron_archive.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/parquet-go/parquet-go"
)

// ParquetPosition is the schema for the Parquet file
type ParquetPosition struct {
	RecordedAt  string  `parquet:"recorded_at"`
	VehicleID   string  `parquet:"vehicle_id"`
	VehicleNum  string  `parquet:"vehicle_num"`
	Route       string  `parquet:"route"`
	TripID      string  `parquet:"trip_id"`
	DirectionID int32   `parquet:"direction_id"`
	Lat         float64 `parquet:"lat"`
	Lon         float64 `parquet:"lon"`
	Speed       float32 `parquet:"speed"`
	Heading     float32 `parquet:"heading"`
}

func getR2Client() (*s3.Client, string) {
	endpoint := os.Getenv("R2_ENDPOINT")
	accessKeyID := os.Getenv("R2_ACCESS_KEY_ID")
	secretAccessKey := os.Getenv("R2_SECRET_ACCESS_KEY")

	if endpoint == "" || accessKeyID == "" || secretAccessKey == "" {
		return nil, ""
	}

	bucket := os.Getenv("R2_BUCKET")
	if bucket == "" {
		bucket = "porto-move"
	}

	client := s3.New(s3.Options{
		BaseEndpoint: &endpoint,
		Region:       "auto",
		Credentials:  credentials.NewStaticCredentialsProvider(accessKeyID, secretAccessKey, ""),
	})

	return client, bucket
}

// runArchivePositions reads yesterday's R2 snapshot files and writes a single
// Parquet archive to positions/YYYY/MM/DD.parquet. No database access needed.
func runArchivePositions(ctx context.Context, r2 *s3.Client, bucket string) error {
	startTime := time.Now()

	now := time.Now().UTC()
	yesterday := time.Date(now.Year(), now.Month(), now.Day()-1, 0, 0, 0, 0, time.UTC)
	dateStr := yesterday.Format("2006-01-02")

	key := fmt.Sprintf("positions/%04d/%02d/%02d.parquet",
		yesterday.Year(), yesterday.Month(), yesterday.Day())

	// Check if already archived (idempotent)
	_, err := r2.HeadObject(ctx, &s3.HeadObjectInput{
		Bucket: &bucket,
		Key:    &key,
	})
	if err == nil {
		log.Printf("[archive] %s already exists — skipping", key)
		return nil
	}

	// List snapshot files for yesterday from R2
	keys, err := listSnapshotKeys(ctx, r2, bucket, dateStr)
	if err != nil {
		return fmt.Errorf("list snapshots: %w", err)
	}
	if len(keys) == 0 {
		log.Printf("[archive] No snapshots for %s", dateStr)
		return nil
	}
	log.Printf("[archive] Reading %d snapshot files for %s", len(keys), dateStr)

	// Read all snapshots and convert to Parquet rows
	var rows []ParquetPosition
	for _, snapKey := range keys {
		out, err := r2.GetObject(ctx, &s3.GetObjectInput{Bucket: &bucket, Key: &snapKey})
		if err != nil {
			log.Printf("[archive] WARNING: failed to fetch %s: %v", snapKey, err)
			continue
		}
		body, err := io.ReadAll(out.Body)
		out.Body.Close()
		if err != nil {
			log.Printf("[archive] WARNING: failed to read %s: %v", snapKey, err)
			continue
		}
		var snap SnapshotFile
		if err := json.Unmarshal(body, &snap); err != nil {
			log.Printf("[archive] WARNING: failed to parse %s: %v", snapKey, err)
			continue
		}
		for _, p := range snap.Positions {
			row := ParquetPosition{
				RecordedAt: snap.RecordedAt,
				VehicleID:  p.VehicleID,
				VehicleNum: p.VehicleNum,
				Route:      p.Route,
				TripID:     p.TripID,
				Lat:        p.Lat,
				Lon:        p.Lon,
			}
			if p.DirectionID != nil {
				row.DirectionID = int32(*p.DirectionID)
			} else {
				row.DirectionID = -1
			}
			if p.Speed != nil {
				row.Speed = float32(*p.Speed)
			} else {
				row.Speed = -1
			}
			// Heading not available in snapshots
			row.Heading = -1
			rows = append(rows, row)
		}
	}

	if len(rows) == 0 {
		log.Printf("[archive] No positions in snapshots for %s", dateStr)
		return nil
	}

	log.Printf("[archive] Writing %d positions to %s", len(rows), key)

	// Write Parquet to buffer
	var buf bytes.Buffer
	writer := parquet.NewGenericWriter[ParquetPosition](&buf)
	if _, err := writer.Write(rows); err != nil {
		return fmt.Errorf("write parquet rows: %w", err)
	}
	if err := writer.Close(); err != nil {
		return fmt.Errorf("close parquet writer: %w", err)
	}

	// Upload to R2
	body := buf.Bytes()
	contentType := "application/vnd.apache.parquet"
	_, err = r2.PutObject(ctx, &s3.PutObjectInput{
		Bucket:      &bucket,
		Key:         &key,
		Body:        bytes.NewReader(body),
		ContentType: &contentType,
		Metadata: map[string]string{
			"rows": fmt.Sprintf("%d", len(rows)),
			"date": dateStr,
		},
	})
	if err != nil {
		return fmt.Errorf("upload to R2: %w", err)
	}

	// Free rows memory
	rows = nil
	_ = rows

	elapsed := time.Since(startTime)
	sizeMB := float64(len(body)) / 1024 / 1024
	log.Printf("[archive] Archived %d positions (%.2f MB) to %s in %s",
		len(body), sizeMB, key, elapsed)
	return nil
}

// listSnapshotKeysForCleanup returns snapshot object keys older than the cutoff.
// It lists all snapshots/ prefixed objects and filters by date directory.
func listSnapshotKeysForCleanup(ctx context.Context, r2 *s3.Client, bucket string, beforeDate time.Time) ([]string, error) {
	prefix := "snapshots/"
	var keys []string
	var continuationToken *string

	for {
		out, err := r2.ListObjectsV2(ctx, &s3.ListObjectsV2Input{
			Bucket:            &bucket,
			Prefix:            &prefix,
			ContinuationToken: continuationToken,
		})
		if err != nil {
			return nil, fmt.Errorf("list R2 objects: %w", err)
		}
		for _, obj := range out.Contents {
			if obj.Key == nil {
				continue
			}
			k := *obj.Key
			// Skip today.json
			if k == "snapshots/today.json" {
				continue
			}
			// Keys look like snapshots/YYYY/MM/DD/HHMMSS.json
			// Extract date from path
			parts := strings.Split(strings.TrimPrefix(k, "snapshots/"), "/")
			if len(parts) < 3 {
				continue
			}
			dateStr := parts[0] + "-" + parts[1] + "-" + parts[2]
			t, err := time.Parse("2006-01-02", dateStr)
			if err != nil {
				continue
			}
			if t.Before(beforeDate) {
				keys = append(keys, k)
			}
		}
		if out.IsTruncated == nil || !*out.IsTruncated {
			break
		}
		continuationToken = out.NextContinuationToken
	}
	return keys, nil
}

File: worker-node/src/cron-archive.ts
/**
 * Cron: Archive yesterday's raw positions as Parquet to Cloudflare R2
 *
 * Runs at 03:30 UTC daily (after aggregation at 03:00).
 * Writes a single Parquet file per day: positions/YYYY/MM/DD.parquet
 * Then the cleanup job (04:00) can safely delete old raw data from Neon.
 */

import {
  S3Client,
  PutObjectCommand,
  HeadObjectCommand,
} from "@aws-sdk/client-s3";
import { parquetWriteBuffer } from "hyparquet-writer";
import { prisma } from "./prisma.js";

// ---------------------------------------------------------------------------
// R2 client (S3-compatible)
// ---------------------------------------------------------------------------

function getR2Client(): S3Client | null {
  const endpoint = process.env.R2_ENDPOINT;
  const accessKeyId = process.env.R2_ACCESS_KEY_ID;
  const secretAccessKey = process.env.R2_SECRET_ACCESS_KEY;

  if (!endpoint || !accessKeyId || !secretAccessKey) {
    console.warn("[archive] R2 not configured — skipping archive");
    return null;
  }

  return new S3Client({
    region: "auto",
    endpoint,
    credentials: { accessKeyId, secretAccessKey },
  });
}

const R2_BUCKET = process.env.R2_BUCKET || "porto-move";

// ---------------------------------------------------------------------------
// Archive pipeline
// ---------------------------------------------------------------------------

export async function runArchivePositions(): Promise<void> {
  const startTime = Date.now();

  const r2 = getR2Client();
  if (!r2) return;

  // Archive yesterday's data
  const now = new Date();
  const yesterday = new Date(now);
  yesterday.setUTCDate(yesterday.getUTCDate() - 1);
  yesterday.setUTCHours(0, 0, 0, 0);
  const today = new Date(yesterday);
  today.setUTCDate(today.getUTCDate() + 1);

  const year = yesterday.getUTCFullYear();
  const month = String(yesterday.getUTCMonth() + 1).padStart(2, "0");
  const day = String(yesterday.getUTCDate()).padStart(2, "0");
  const key = `positions/${year}/${month}/${day}.parquet`;

  // Check if already archived (idempotent)
  try {
    await r2.send(new HeadObjectCommand({ Bucket: R2_BUCKET, Key: key }));
    console.log(`[archive] ${key} already exists — skipping`);
    return;
  } catch {
    // Object doesn't exist — proceed with archive
  }

  // Fetch all positions for yesterday in batches to limit memory
  const BATCH_SIZE = 50_000;
  let offset = 0;
  const allRecordedAt: string[] = [];
  const allVehicleId: string[] = [];
  const allVehicleNum: (string | null)[] = [];
  const allRoute: (string | null)[] = [];
  const allTripId: (string | null)[] = [];
  const allDirectionId: (number | null)[] = [];
  const allLat: number[] = [];
  const allLon: number[] = [];
  const allSpeed: (number | null)[] = [];
  const allHeading: (number | null)[] = [];

  while (true) {
    const batch = await prisma.busPositionLog.findMany({
      where: { recordedAt: { gte: yesterday, lt: today } },
      orderBy: { recordedAt: "asc" },
      skip: offset,
      take: BATCH_SIZE,
    });

    if (batch.length === 0) break;

    for (const p of batch) {
      allRecordedAt.push(p.recordedAt.toISOString());
      allVehicleId.push(p.vehicleId);
      allVehicleNum.push(p.vehicleNum);
      allRoute.push(p.route);
      allTripId.push(p.tripId);
      allDirectionId.push(p.directionId);
      allLat.push(p.lat);
      allLon.push(p.lon);
      allSpeed.push(p.speed);
      allHeading.push(p.heading);
    }

    offset += batch.length;
    if (batch.length < BATCH_SIZE) break;
  }

  const totalRows = allRecordedAt.length;
  if (totalRows === 0) {
    console.log(`[archive] No positions for ${year}-${month}-${day}`);
    return;
  }

  console.log(`[archive] Writing ${totalRows} positions to ${key}`);

  // Write Parquet in memory
  const parquetBytes = parquetWriteBuffer({
    columnData: [
      { name: "recorded_at", data: allRecordedAt, type: "STRING" },
      { name: "vehicle_id", data: allVehicleId, type: "STRING" },
      {
        name: "vehicle_num",
        data: allVehicleNum.map((v) => v ?? ""),
        type: "STRING",
      },
      {
        name: "route",
        data: allRoute.map((v) => v ?? ""),
        type: "STRING",
      },
      {
        name: "trip_id",
        data: allTripId.map((v) => v ?? ""),
        type: "STRING",
      },
      {
        name: "direction_id",
        data: allDirectionId.map((v) => v ?? -1),
        type: "INT32",
      },
      { name: "lat", data: allLat, type: "DOUBLE" },
      { name: "lon", data: allLon, type: "DOUBLE" },
      {
        name: "speed",
        data: allSpeed.map((v) => v ?? -1),
        type: "FLOAT",
      },
      {
        name: "heading",
        data: allHeading.map((v) => v ?? -1),
        type: "FLOAT",
      },
    ],
  });

  // Upload to R2
  const body = Buffer.from(parquetBytes);
  await r2.send(
    new PutObjectCommand({
      Bucket: R2_BUCKET,
      Key: key,
      Body: body,
      ContentType: "application/vnd.apache.parquet",
      Metadata: {
        rows: String(totalRows),
        date: `${year}-${month}-${day}`,
      },
    })
  );

  const elapsed = Date.now() - startTime;
  const sizeMB = (body.byteLength / 1024 / 1024).toFixed(2);
  console.log(
    `[archive] Archived ${totalRows} positions (${sizeMB} MB) to ${key} in ${elapsed}ms`
  );
}
